pkg/metrics/stats_reporter.go

```diff
@@ -0,0 +1,103 @@
+package metrics
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+)
+
+const (
+	instrumentationName       = "keyvaultkms"
+	errorTypeKey              = "error_type"
+	statusTypeKey             = "status"
+	errorStatusTypeValue      = "error"
+	successStatusTypeValue    = "success"
+	operationTypeKey          = "operation"
+	encryptOperationTypeValue = "encrypt"
+	decryptOperationTypeValue = "decrypt"
+)
+
+var (
+	totalRequest    metric.Int64Counter
+	requestDuration metric.Float64ValueRecorder
+)
+
+type reporter struct {
+	meter metric.Meter
+}
+
+// StatsReporter reports metrics
+type StatsReporter interface {
+	ReportEncryptCountMetric(ctx context.Context)
+	ReportDecryptCountMetric(ctx context.Context)
+	ReportEncryptErrorCountMetric(ctx context.Context, errorMessage string)
+	ReportDecryptErrorCountMetric(ctx context.Context, errorMessage string)
+	ReportEncryptDurationMetric(ctx context.Context, duration float64)
+	ReportDecryptDurationMetric(ctx context.Context, duration float64)
```

Comment:
I think you can combine all of these

Reply:
I know you can also have 1 metric that can have 2 views: count and histogram. so that might apply here too

Reply:
Make sense.Histogram gives distribution along with sum and count metric. Here is updated metric output.

```
# HELP kms_request Distribution of how long it took for an operation
# TYPE kms_request histogram
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.1"} 59
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.2"} 68
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.3"} 72
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.4"} 107
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.5"} 144
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="1"} 190
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="1.5"} 191
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="2"} 191
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="2.5"} 200
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="3"} 200
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="5"} 200
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="10"} 200
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="15"} 200
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="30"} 200
kms_request_bucket{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="+Inf"} 200
kms_request_sum{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0"} 84.844736613
kms_request_count{operation="decrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0"} 200
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.1"} 0
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.2"} 0
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.3"} 0
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.4"} 0
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="0.5"} 0
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="1"} 0
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="1.5"} 121
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="2"} 141
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="2.5"} 147
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="3"} 180
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="5"} 200
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="10"} 200
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="15"} 200
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="30"} 200
kms_request_bucket{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0",le="+Inf"} 200
kms_request_sum{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0"} 371.51619479100015
kms_request_count{operation="encrypt",service_name="unknown_service:__debug_bin",status="success",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.20.0"} 200
```